src/util/imagefiledata.cpp

```diff
@@ -0,0 +1,82 @@
+#include "util/imagefiledata.h"
+
+#include <QBuffer>
+#include <QByteArray>
+#include <QFile>
+#include <QImage>
+#include <QImageReader>
+
+#include "util/assert.h"
+
+ImageFileData::ImageFileData(
+        const QByteArray& coverArtBytes, const QByteArray& coverArtFormat)
+        : QImage(QImage::fromData(coverArtBytes)),
+          m_coverArtBytes(coverArtBytes) {
+    m_coverArtFormat = coverArtFormat.isEmpty() ? readFormatFrom(m_coverArtBytes) : coverArtFormat;
+}
+
+//static
+ImageFileData ImageFileData::fromFilePath(const QString& coverArtFilePath) {
+    QFile coverArtFile = QFile(coverArtFilePath);
+    if (!coverArtFile.open(QIODevice::ReadOnly)) {
+        qWarning() << "Failed to open cover art file";
+        return ImageFileData(QByteArray{}, QByteArray{});
+    }
+    QByteArray coverArtBytes = coverArtFile.readAll();
+    QByteArray coverArtFormat = readFormatFrom(coverArtBytes);
+    return ImageFileData(coverArtBytes, coverArtFormat);
+}
+
+//static
+QByteArray ImageFileData::readFormatFrom(const QByteArray& coverArtBytes) {
+    QByteArray bytes = coverArtBytes;
+    QBuffer b = QBuffer(&bytes);
+    if (!b.open(QIODevice::ReadOnly)) {
+        qWarning() << "Failed to open buffer";
+        return QByteArray{};
+    }
+    auto reader = QImageReader(&b);
+    QByteArray coverArtFormat = reader.format();
+    VERIFY_OR_DEBUG_ASSERT(!coverArtFormat.isEmpty()) {
+        return QByteArray{};
+    }
+    return coverArtFormat;
+}
+
+bool ImageFileData::saveFile(const QString& coverArtAbsoluteFilePath) const {
+    if (m_coverArtBytes.isEmpty()) {
+        return save(coverArtAbsoluteFilePath);
```

Comment:
I think save() will fail if the extension is not detected: 
https://github.com/qt/qtbase/blob/9d2cc4dd766ca6538e17040b6ac845ed880ab0fe/tests/auto/gui/image/qimagewriter/tst_qimagewriter.cpp#L373

Below we write the m_coverArtBytes unconditionally. 
I think we should also fail if the extension in coverArtAbsoluteFilePath does not match the m_coverArtFormat.